core/include/bitarithm.h

```diff
@@ -164,6 +164,45 @@ static inline unsigned bitarithm_lsb(unsigned v)
 }
 #endif
 
+/**
+ * @brief   Used for iterating over the bits in @state.
+ *          Returns the index of a set bit in @p state, returns @p state with that bit cleared.
+ *
+ * @warning This is a low-level helper function, arguments are not checked.
+ *          It is intended to iterate over a bit map until all bits are cleared.
+ *          Whether it starts with the highest or lowest bit will depend on what is fastest
+ *          on the given hardware.
+ *
+ * @warning @p state must not be zero.
+ *
+ * @param[in]   state   Bit Map with at least one bit set
+ * @param[out]  index   Index of the first set bit. Must be initialized with 0 before the
+ *                      first call to this function, must not be modified between subsequent
+ *                      calls.
+ *
+ * @return      new state value - must be treated as opaque value
+ *
+ */
+static inline unsigned bitarithm_test_and_clear(unsigned state, uint8_t *index)
+{
+#if defined(BITARITHM_HAS_CLZ)
+    *index = 8 * sizeof(state) - __builtin_clz(state) - 1;
+    return state & ~(1 << *index);
+#elif defined(BITARITHM_LSB_LOOKUP)
+    /* Source: http://graphics.stanford.edu/~seander/bithacks.html#ZerosOnRightMultLookup */
+    extern const uint8_t MultiplyDeBruijnBitPosition[32];
+    uint32_t least_bit = state & -state;
```

Comment:
Uncrustify is complaining:

```suggestion
    uint32_t least_bit = state & - state;
```

Reply:
That whitespace makes it more confusing imho.
`-state` is negative state `- state` looks like it would be subtracted from something.

Reply:
I don't disagree with you, but I don't like to have a permanent uncrustify warning around in `core/`. @kaspar030 Is there a way to tune uncrustify to agree with us here?

Reply:
Have you tried `state & (-state)`? Maybe this makes `uncrustify` happy, without sacrificing readability. (`state & - state` would take me minutes to even understand why it compiles, as it is common sense to not add spaces after unary operators to distinguish them from binary operators.)

Reply:
It's the same in `bitarithm_lsb()` and Travis doesn't seem to care - maybe similar to #14514

Reply:
`state & (-state)` seems to work:

```
$ cat test.c 
int foo(uint32_t bar)
{
    return bar & (-bar);
}

$ uncrustify -c uncrustify-riot.cfg  -f test.c 
Parsing: test.c as language C
int foo(uint32_t bar)
{
    return bar & (-bar);
}
```

Reply:
(Some people might find `state & (-state)` even more readable than `state & -state`.)

Reply:
`state &- state` is the obvious way to go :wink:

Reply:
What version of `uncrustify` is that?

```
$ cat test.c       
int foo(uint32_t bar)
{
    return bar & -bar;
}

$ uncrustify -c uncrustify-riot.cfg -f test.c     
Parsing: test.c as language C
int foo(uint32_t bar)
{
    return bar & -bar;
}

$ uncrustify -v
Uncrustify-0.69.0_f

```

Reply:
On version `0.71.0_f` `bar & -bar` is also not touched. (I didn't check this before, sorry.) So I guess we can consider this a bug in uncrustify that was fixed upstream, but the fix was not yet shipped to docker image in the CI?

Reply:
Ok, so this should sort itself out eventually.